src/mon/MDSMonitor.cc

```diff
@@ -1759,18 +1715,32 @@ int MDSMonitor::print_nodes(Formatter *f)
 
 /**
  * If a cluster is undersized (with respect to max_mds), then
- * attempt to find daemons to grow it.
+ * attempt to find daemons to grow it. If the cluster is oversized
+ * (with respect to max_mds) then shrink it by stopping its highest rank.
  */
-bool MDSMonitor::maybe_expand_cluster(std::shared_ptr<Filesystem> fs)
+bool MDSMonitor::maybe_resize_cluster(std::shared_ptr<Filesystem> fs)
 {
-  bool do_propose = false;
+  if (fs->mds_map.test_flag(CEPH_MDSMAP_NOT_JOINABLE)) {
+    return false;
+  }
 
-  if (fs->mds_map.test_flag(CEPH_MDSMAP_DOWN)) {
-    return do_propose;
+  int in = fs->mds_map.get_num_in_mds();
+  int max = fs->mds_map.get_max_mds();
+
+  dout(20) << __func__ << " in " << in << " max " << max << dendl;
+
+  if (fs->mds_map.is_degraded()) {
+    dout(5) << "not resizing degraded MDS cluster "
+	         << fs->mds_map.fs_name << dendl;
+    return false;
+  }
+
+  if (fs->mds_map.get_num_mds(CEPH_MDS_STATE_STOPPING)) {
+    dout(5) << "An MDS for " << fs->mds_map.fs_name
+	         << " is stopping; waiting to resize" << dendl;
```

Comment:
I don't see us waiting anywhere. Is this by design? Are we okay to fall-through to the next set of instructions? Maybe return here instead?

Reply:
Ya that's a bug I've fixed in my testing. It'll get pushed soon.